### PR TITLE
Tour around Weld 3

### PR DESCRIPTION
**Pull request by manovotn**

### CODE REVIEW

**Comment by manovotn** [comment]
@mkouba feel free to take over and re-design this to your liking.

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+desc: A summary of CDI 2.0 features coated with Weld and sprinkled with code samples
+tags: [release]
+---
+:linkattrs:
+
+This post briefly describes all the main CDI 2.0 features and elaborates on Weld-specific features we added on top of that.
+It is not intended as a deep-dive but rather to give you the overall idea of what is going on and what can the new release offer.
+So, enough talk, let's get the show on the road!
+
+=== Async Events and Notification Options
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: This should be h2, ie. `==`

**Comment by mkouba** [comment]
I would add TOC as we did in tips...

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+Here is how to fire such and event and how to observe it:
+
+[ source, java ]
+----
+// nothing new here, plain old Event is used
+@Inject Event<Payload> event;
+
+public void sendAsyncEvent() {
+    // we use a new fireAsync method for asynchronous events
+    CompletionStage<Void> eventResult = event.fireAsync(new Payload()).thenAccept(...);
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: This should be `CompletionStage<Payload> eventResult =...`

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+From CDI perspective, there is currently only one standardized option and that enables usage of custom executors which will be used for notifications.
+But the API was designed in a generic way so that implementation can enrich it with its own options.
+Weld currently offers two more options - timeout and parallel execution - both of which are well described in link:http://docs.jboss.org/weld/reference/3.0.0.Final/en-US/html/events.html#_notification_options[Weld docs, window="_blank"].
+And here is a snippet showing it in action:
+
+[ source, java ]
+----
+public void sendAsyncEventWithOptions() {
+    // we use a secondary version of fireAsync method with NotificationOptions parameter
+    event.fireAsync(new Payload(), NotificationOptions.of("weld.async.notification.timeout", "2000"));
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: We do support string values here but I would rather use integer here, ie. `2000`.

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    // we use a secondary version of fireAsync method with NotificationOptions parameter
+    event.fireAsync(new Payload(), NotificationOptions.of("weld.async.notification.timeout", "2000")); 
+}
+----
+
+
+=== Configurators API
+
+Another of the key features are link:http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#configurators[`Configurators`, window="blank"].
+They are there to ease our work when doing small tweaks in extensions.
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: `They are here` sounds better to me... but it's probably just a matter of taste ;-)

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+Here is a snippet showing how easy it can be to register a new bean in `AfterBeanDiscovery` using configurator.
+You don't a new class implementing `Bean<X>` anymore:
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: Missing verb...

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+[ source, java ]
+----
+public void addABean(@Observes AfterBeanDiscovery event) {
+    // get an instance of BeanConfigurator
+    event.addBean()
+      // set the desired data
+      .beanClass(Foo.class)
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: `beanClass()` is optional, we should rather set bean types, e.g. `.types(Foo.class)`.

And this snippet is missing a callback to create a bean instance.

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+      .scope(RequestScoped.class)
+      .addQualifier(Custom.CustomLiteral.INSTANCE);
+}
+----
+
+Following configurators were added:
+
+* AnnotatedTypeConfigurator
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: We should use literals here...

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+* InjectionPointConfigurator
+* BeanAttributesConfigurator
+* BeanConfigurator
+* ObserverMethodConfigurator
+
+On top of that, Weld adds one additional configurator - link:http://docs.jboss.org/weld/reference/latest-master/en-US/html_single/#_weld_enriched_container_lifecycle_events['InterceptorConfigurator`, window="_blank"].
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: 'InterceptorConfigurator -> replace the apostrophe with backtick.

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        })).build();
+
+try (WeldContainer container = new Weld().addExtension(testExtension).initialize()) {
+    // Use the container...
+}
+----
+
+What we just did was to create an extension with two container lifecycle observer methods.
+It follows a familiar builder pattern and the body of the observer methods is specified as lambda expression.
+Before booting Se container, we register this extension as we would any other.
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: I would add a link to the blogpost we have...

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+Then we `configure()` it, which returns an `AnnotatedTypeConfigurator` allowing us to add interceptor binding.
+Finally, we invoke `InterceptionFactory.createInterceptedInstance()` which takes a the object we produce as a parameter.
+
+You can read more about this in link:http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#interception_factory[this CDI spec chapter, window="_blank"].
+
+=== Trimmed Bean Archives
+
+Last but not least feature we will mention as so called 'trimmed' bean archives.
+You can mark an explicit bean archive as trimmed in `beans.xml` by adding the `</trim>` element.
+Such bean archive will perform an annotated type discovery as with `bean-discovery-mode="all"` but all types that don't have a bean defining annotation are then removed from the set of discovered types.
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: We should probably mention http://docs.jboss.org/weld/reference/latest-master/en-US/html/configure.html#veto-types-without-bean-defining-annotation and the difference...

File: news/2017-05-19-tour-around-weld-3.asciidoc (hunk 7 above)
Comment (manovotn) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: I don't understand what do you mean by literals here.

Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: Code literals, backticks, producing `<code></code>`.

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 11]
```diff
@@ -0,0 +1,207 @@
+---
+layout: news
+title: Tour around Weld 3
+author: Matej Novotny
+priority: 1
+change_frequency: weekly
+excerpt: Tour around the new stuff in Weld 3
+desc: A summary of CDI 2.0 features coated with Weld and sprinkled with code samples
+tags: [tour,weld3,noteworthy]
```
Comment (manovotn) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: Not sure what kind of tags would we like for this article?

Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: I think `cdi2` and `weld3` should be enough.

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+    // we use a new fireAsync method for asynchronous events
+    CompletionStage<Payload> eventResult = event.fireAsync(new Payload()).thenAccept(...); 
+}
+
+public void asyncObserver (@ObservesAsync Payload payload){ … }
+----
+
+Few things to note here - first of all, the return value of `fireAsync` is `CompletionStage` allowing you to chain the work for when the event is done.
+You can follow-up with more tasks, react on exceptional return value and so on.
+Another noticeable detail is that observer which are to be notified have to have `@ObservesAsync` annotation.
```
Comment (tremes) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: Missing `s` for observers

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+}
+----
+
+Following configurators were added:
+
+* `AnnotatedTypeConfigurator`
+* `InjectionPointConfigurator`
+* `BeanAttributesConfigurator`
+* `BeanConfigurator`
+* `ObserverMethodConfigurator`
```
Comment (tremes) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: Minor note - there is also `ProducerConfigurator`

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+Then we `configure()` it, which returns an `AnnotatedTypeConfigurator` allowing us to add interceptor binding.
+Finally, we invoke `InterceptionFactory.createInterceptedInstance()` which takes a the object we produce as a parameter.
+
+You can read more about this in link:http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#interception_factory[this CDI spec chapter, window="_blank"].
+
+== Trimmed Bean Archives
+
+Last but not least feature we will mention are so called 'trimmed' bean archives.
+You can mark an explicit bean archive as trimmed in `beans.xml` by adding the `</trim>` element.
+Such bean archive will perform an annotated type discovery as with `bean-discovery-mode="all"` but all types that don't have a bean defining annotation are then removed from the set of discovered types.
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: > all types that don't have a bean defining annotation **or any scope annotation...**

File: news/2017-05-19-tour-around-weld-3.asciidoc [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+
+Even in this case, Weld allows you to go one step further and link:http://docs.jboss.org/weld/reference/latest-master/en-US/html/configure.html#veto-types-without-bean-defining-annotation[veto types based on regular expression].
+It works in very much the same way.
```
Comment (mkouba) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: Well, the difference is that this is applied globally - Weld extension is processing all types from all bean archives.

File: news/2017-05-19-tour-around-weld-3.asciidoc (hunk 13 above)
Comment (manovotn) [inline news/2017-05-19-tour-around-weld-3.asciidoc]: Good point, added.

**Pull request merged**